File: src/main/java/com/milaboratory/mixcr/cli/CommandAlignmentsStats.java
/*
 * Copyright (c) 2014-2021, MiLaboratories Inc. All Rights Reserved
 *
 * BEFORE DOWNLOADING AND/OR USING THE SOFTWARE, WE STRONGLY ADVISE
 * AND ASK YOU TO READ CAREFULLY LICENSE AGREEMENT AT:
 *
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 */
package com.milaboratory.mixcr.cli;

import cc.redberry.pipe.CUtils;
import cc.redberry.pipe.VoidProcessor;
import com.milaboratory.mixcr.basictypes.VDJCAlignments;
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsReader;
import com.milaboratory.mixcr.info.AlignmentInfoCollector;
import com.milaboratory.mixcr.info.GeneFeatureCoverageCollector;
import com.milaboratory.mixcr.info.ReferencePointCoverageCollector;
import com.milaboratory.util.SmartProgressReporter;
import io.repseq.core.GeneFeature;
import io.repseq.core.ReferencePoint;
import picocli.CommandLine.Command;
import picocli.CommandLine.Parameters;

import java.io.BufferedOutputStream;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.util.Collections;
import java.util.List;

import static io.repseq.core.GeneFeature.*;
import static io.repseq.core.ReferencePoint.*;

@Command(name = "alignmentsStat",
        sortOptions = true,
        hidden = true,
        separator = " ",
        description = "Alignments statistics.")
public class CommandAlignmentsStats extends ACommandMiXCR {
    @Parameters(index = "0", description = "input_file.vdjca")
    public String in;
    @Parameters(index = "1", description = "[output.txt]", arity = "0..1")
    public String out = null;

    @Override
    protected List<String> getInputFiles() {
        return Collections.singletonList(in);
    }

    @Override
    protected List<String> getOutputFiles() {
        return out == null ? Collections.emptyList() : Collections.singletonList(out);
    }


    @Override
    public void run0() throws Exception {
        long[] geneFeatureCounters = new long[targetFeatures.length];
        AlignmentInfoCollector[] collectors = new AlignmentInfoCollector[targetFeatures.length + targetReferencePoints.length];

        int i = 0;

        for (GeneFeature targetFeature : targetFeatures)
            collectors[i++] = new GeneFeatureCoverageCollector(targetFeature);

        for (ReferencePoint targetReferencePoint : targetReferencePoints)
            collectors[i++] = new ReferencePointCoverageCollector(targetReferencePoint, 40, 40);

        final Collector collector = new Collector(collectors);

        try (VDJCAlignmentsReader reader = new VDJCAlignmentsReader(in);
             PrintStream output = out == null ? System.out :
                     new PrintStream(new BufferedOutputStream(new FileOutputStream(out), 32768))
        ) {
            SmartProgressReporter.startProgressReport("Analysis", reader);
            CUtils.processAllInParallel(reader, collector, Math.min(4, Runtime.getRuntime().availableProcessors()));
            collector.end();

            if (output == System.out)
                output.println();

            collector.write(output);
        }
    }

    private static final GeneFeature[] targetFeatures = {
            V5UTR,
            new GeneFeature(L1Begin, -20, 0),
            L1, VIntron, L2, FR1, CDR1, FR2, CDR2, FR3, CDR3, FR4,
            new GeneFeature(FR4, 0, -3)
    };

    private static final ReferencePoint[] targetReferencePoints = {
            L1Begin, L1End, L2Begin, FR1Begin, CDR1Begin, FR2Begin, CDR2Begin, FR3Begin, CDR3Begin,
            FR4Begin, FR4End
    };

    private static class Collector implements VoidProcessor<VDJCAlignments> {
        final AlignmentInfoCollector[] collectors;

        public Collector(AlignmentInfoCollector... collectors) {
            this.collectors = collectors;
        }

        @Override
        public void process(VDJCAlignments input) {
            for (AlignmentInfoCollector collector : collectors)
                collector.put(input);
        }

        public void end() {
            for (AlignmentInfoCollector collector : collectors)
                collector.end();
        }

        public void write(PrintStream writer) {
            for (AlignmentInfoCollector collector : collectors)
                collector.writeResult(writer);
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandAlignmentsStats.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import cc.redberry.pipe.CUtils
import cc.redberry.pipe.VoidProcessor
import com.milaboratory.mixcr.basictypes.VDJCAlignments
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsReader
import com.milaboratory.mixcr.info.AlignmentInfoCollector
import com.milaboratory.mixcr.info.GeneFeatureCoverageCollector
import com.milaboratory.mixcr.info.ReferencePointCoverageCollector
import com.milaboratory.util.SmartProgressReporter
import io.repseq.core.GeneFeature
import io.repseq.core.ReferencePoint
import picocli.CommandLine.Command
import picocli.CommandLine.Parameters
import java.io.BufferedOutputStream
import java.io.FileOutputStream
import java.io.PrintStream
import java.nio.file.Path
import kotlin.math.min

private val targetFeatures = arrayOf(
    GeneFeature.V5UTR,
    GeneFeature(ReferencePoint.L1Begin, -20, 0),
    GeneFeature.L1,
    GeneFeature.VIntron,
    GeneFeature.L2,
    GeneFeature.FR1,
    GeneFeature.CDR1,
    GeneFeature.FR2,
    GeneFeature.CDR2,
    GeneFeature.FR3,
    GeneFeature.CDR3,
    GeneFeature.FR4,
    GeneFeature(GeneFeature.FR4, 0, -3)
)
private val targetReferencePoints = arrayOf(
    ReferencePoint.L1Begin,
    ReferencePoint.L1End,
    ReferencePoint.L2Begin,
    ReferencePoint.FR1Begin,
    ReferencePoint.CDR1Begin,
    ReferencePoint.FR2Begin,
    ReferencePoint.CDR2Begin,
    ReferencePoint.FR3Begin,
    ReferencePoint.CDR3Begin,
    ReferencePoint.FR4Begin,
    ReferencePoint.FR4End
)

@Command(
    hidden = true,
    description = ["Alignments statistics."]
)
class CommandAlignmentsStats : MiXCRCommandWithOutputs() {
    @Parameters(index = "0", description = ["input_file.vdjca"])
    lateinit var input: Path

    @Parameters(index = "1", description = ["[output.txt]"], arity = "0..1")
    var out: Path? = null

    override val inputFiles
        get() = listOf(input)

    override val outputFiles
        get() = listOfNotNull(out)

    override fun run0() {
        val collectors = targetFeatures.map { GeneFeatureCoverageCollector(it) } +
                targetReferencePoints.map { ReferencePointCoverageCollector(it, 40, 40) }

        val collector = Collector(collectors)
        VDJCAlignmentsReader(input).use { reader ->
            (out?.let { PrintStream(BufferedOutputStream(FileOutputStream(it.toFile()), 32768)) }
                ?: System.out).use { output ->
                SmartProgressReporter.startProgressReport("Analysis", reader)
                CUtils.processAllInParallel(reader, collector, min(4, Runtime.getRuntime().availableProcessors()))
                collector.end()
                if (output === System.out) output.println()
                collector.write(output)
            }
        }
    }

    private class Collector(val collectors: List<AlignmentInfoCollector>) : VoidProcessor<VDJCAlignments> {
        override fun process(input: VDJCAlignments) = collectors.forEach { collector ->
            collector.put(input)
        }

        fun end() = collectors.forEach { collector ->
            collector.end()
        }

        fun write(writer: PrintStream) = collectors.forEach { collector ->
            collector.writeResult(writer)
        }
    }
}
